Status check

jsliacan opened the pull request:
**Fixes:** Issue #1737 

## Solution

The entire `crc status` check is hidden in one function. If looking for `Stopped` status, then exit with success as soon as status returns `Stopped` for the first time. If looking for `Running` status, then exit with success only if 3 consecutive probes (at 1min intervals) return `Running` cluster status. The probing is set to 15x1min.

Second commit updates the language in Feature files.

## Testing

1. `make integration` succeeds

praveenkumar commented on test/integration/crcsuite/crcsuite.go:
```diff
@@ -247,6 +249,48 @@ func CheckOutputMatchWithRetry(retryCount int, retryTime string, command string,
+		numConsecutive = 1
+		expression = `.*Stopped.*`
+	} else {
+		numConsecutive = 3
```
So this number is present to have consecutive execute the crc status and see if have same output?

jsliacan replied:
Yes. Exactly. In particular, I assume that if OCP cluster has been Running for 3mins in a row, then it's "stable" for the rest of the Feature (in particular, this check avoids "Running -> Not reachable -> Starting" pitfall right after start.

praveenkumar commented on test/integration/crcsuite/crcsuite.go:
```diff
@@ -247,6 +249,48 @@ func CheckOutputMatchWithRetry(retryCount int, retryTime string, command string,
+// CheckCRCStatus checks that output of status command
+// matches given regex number of consecutive times
+func CheckCRCStatus(state string) error {
+
```
unwanted new line

jsliacan replied:
I'm not sure if it's unwanted (gofmt doesn't mind and the same newlines are in all other functions). But I don't mind, so I removed it. I will go through these files later and also give comments to all exported functions.

praveenkumar commented on test/integration/crcsuite/crcsuite.go:
```diff
@@ -247,6 +249,48 @@ func CheckOutputMatchWithRetry(retryCount int, retryTime string, command string,
+		time.Sleep(retryDuration)
+	}
+
+	return fmt.Errorf("did not get enough consecutive matches")
```
We should also mention how much matches we wanted and how much is present.

praveenkumar commented on test/integration/features/basic.feature:
```diff
@@ -21,18 +21,18 @@ Feature: Basic test
         And stdout should contain "status"
         And stdout should contain "Flags:"
-        And stdout should contain 
+        And stdout should contain
```
This shouldn't be part of this commit.

jsliacan replied:
OK, will address all of these.

gbraad replied:
ETA?

praveenkumar commented on test/integration/features/basic.feature:
```diff
@@ -21,18 +21,18 @@ Feature: Basic test
+        Then stderr should contain
+            """
+            Machine 'crc' does not exist. Use 'crc start' to create it
+            """
```
This shouldn't be part of this commit.

praveenkumar commented on test/integration/features/cert_rotation.feature:
```diff
@@ -1,27 +1,27 @@
+  Scenario: CRC delete and cleanup
+    When executing "crc delete -f" succeeds
+    Then stdout should contain "Deleted the OpenShift cluster"
+    When executing "crc cleanup" succeeds
```
Formatting should be done is different commit message

praveenkumar commented on test/integration/features/story_marketplace.feature:
```diff
@@ -68,7 +68,7 @@ Feature:
         Then with up to "10" retries with wait period of "30s" command "oc get pods" output matches "(?s)(.*example-[a-z0-9]* *1/1 *Running.*){3}"
         But with up to "10" retries with wait period of "30s" command "oc get pods" output does not match "(?s)(.*example-[a-z0-9]* *1/1 *Running.*){4}"
-        
+
```
This shouldn't be part of this commit.

cfergeau commented on test/integration/features/proxy.feature:
```diff
@@ -1,9 +1,8 @@
-  The user tries to use crc behind proxy. he/she expects the
-  crc start happen successfully and able to deploy the app and check its
-  accessibility.
+  User starts CRC behind proxy. They expect a successful start
```
`behind a proxy` maybe?

cfergeau commented on test/integration/features/story_health.feature:
```diff
@@ -1,8 +1,10 @@
+    User goes through a "day in the life of" CRC. They set-up
+    and start CRC. They then create a project and deploy an app.
+    They check on the app and delete the project. They stop and
+    delete CRC.
```
Imo it's weird to talk about "deleting CRC", it's the CRC VM or instance which is being acted on.

cfergeau commented on test/integration/features/story_marketplace.feature:
```diff
@@ -1,7 +1,8 @@
-    admin tasks.
+Feature: Operator from marketplace
+
+    User installs OpenShift operator from OperatorHub and uses it
```
`an OpenShift operator` ?

cfergeau commented on test/integration/crcsuite/crcsuite.go:
```diff
@@ -247,6 +249,48 @@ func CheckOutputMatchWithRetry(retryCount int, retryTime string, command string,
+// matches given regex number of consecutive times
+func CheckCRCStatus(state string) error {
+
+	retryDuration, _ := time.ParseDuration("1m")
```
I think it can just be `retryDuration := 1 * time.Minute`

cfergeau commented on test/integration/crcsuite/crcsuite.go:
```diff
@@ -247,6 +249,48 @@ func CheckOutputMatchWithRetry(retryCount int, retryTime string, command string,
+	retryDuration, _ := time.ParseDuration("1m")
+	retryCount := 15
+	cmd := "crc status --log-level debug"
+	expression := ``
```
You can init with `""`

cfergeau commented on test/integration/crcsuite/crcsuite.go:
```diff
@@ -247,6 +249,48 @@ func CheckOutputMatchWithRetry(retryCount int, retryTime string, command string,
+
+	if state == "Stopped" {
+		numConsecutive = 1
+		expression = `.*Stopped.*`
```
Here you can also use `".*Stopped.*"` I think.

cfergeau commented on test/integration/crcsuite/crcsuite.go:
```diff
@@ -247,6 +249,48 @@ func CheckOutputMatchWithRetry(retryCount int, retryTime string, command string,
+	} else {
+		numConsecutive = 3
+		expression = `.*Running \(v\d+\.\d+\.\d+.*\).*`
+	}
```
And here the backticks are useful to have because of the `\`

jsliacan replied:
I used backticks everywhere because I had to use them here. Will change other places to double quotes.

cfergeau commented on test/integration/crcsuite/crcsuite.go:
```diff
@@ -54,6 +54,8 @@ func FeatureContext(s *godog.Suite) {
 		CheckHTTPResponseWithRetry)
 	s.Step(`^with up to "(\d+)" retries with wait period of "(\d*(?:ms|s|m))" command "(.*)" output (should match|matches|should not match|does not match) "(.*)"$`,
 		CheckOutputMatchWithRetry)
+	s.Step(`^checking that CRC status is (Running|Stopped)$`,
```
nicer to read as `checking that CRC is (running|stopped)$` ?

jsliacan replied:
True. Will address other comments as well. Thanks!

cfergeau reviewed (commented):
Couple of comments, but feel free to ignore :)

jsliacan commented:
@praveenkumar @cfergeau , I addressed all comments (I hope!). Kept 1st commit as the main one. Second improves Feature descriptions and such, and I moved all formatting changes to a new (3rd) commit.

gbraad reviewed (changes requested):
whitespaces/indentations are all over the place

jsliacan commented:
> whitespaces/indentations are all over the place

@gbraad , the purpose of the 3rd commit was to clean that up (when different people write Cucumber code, it indents with 2/4 spaces (even 3!), sometimes with tabs, etc.). Now the Features should be pretty consistent. Any examples of bad indentation?
